pkg/controllers/corecluster/corecluster.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	clusterCopy := cluster.DeepCopy()
+
+	// Check if the Control Plane Initialized condition is set on an already existing cluster.
+	// If it's not, set it.
+	if conditions.Get(cluster, clusterv1.ControlPlaneInitializedCondition) == nil {
+		conditions.MarkTrue(cluster, clusterv1.ControlPlaneInitializedCondition)
+
+		patch := client.MergeFrom(clusterCopy)
+
+		isRequired, err := util.IsPatchRequired(cluster, patch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to check if patch required: %w", err)
+		}
+
+		if isRequired {
+			if err := r.Status().Patch(ctx, cluster, patch); err != nil {
+				return ctrl.Result{}, fmt.Errorf("unable to update core cluster status: %w", err)
+			}
+		}
+
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+	}
+
+	defer logger.V(1).Info("Finished reconciling core cluster")
+
+	return ctrl.Result{}, nil
+}
+
+// CreateCluster creates a cluster with the given name and returns the cluster object.
+func (r *CoreClusterReconciler) ensureCoreCluster(ctx context.Context, clusterName, infraClusterKind string, logger logr.Logger) (*clusterv1.Cluster, error) {
+	cluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: CAPINamespace,
+		}}
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(cluster), cluster); err != nil && !errors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get core cluster: %w", err)
+	} else if err == nil {
+		return cluster, nil
+	}
+
+	logger.Info(fmt.Sprintf("Core Cluster %s/%s does not exist, creating it", CAPINamespace, clusterName))
+
+	apiURL, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiURL: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiURL.Port(), 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiURL port: %w", err)
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return nil, fmt.Errorf("infrastructure PlatformStatus should not be nil: %w", err)
+	}
+
+	cluster = &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: CAPINamespace,
+		},
+		Spec: clusterv1.ClusterSpec{
+			InfrastructureRef: &corev1.ObjectReference{
+				APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+				Kind:       infraClusterKind,
+				Name:       clusterName,
+				Namespace:  CAPINamespace,
+			},
+			ControlPlaneEndpoint: clusterv1.APIEndpoint{
+				Host: apiURL.Hostname(),
+				Port: int32(port),
+			},
+		},
+	}
+
+	clusterCopy := cluster.DeepCopy()
+
+	conditions.MarkTrue(cluster, clusterv1.ControlPlaneInitializedCondition)
```

Comment:
I think this whole chunk setting the available condition may be nicer in it's own function

Reply:
(I've done that now)